File: aula-01/java/src/main/java/com/github/alicefng/cs/aula1/domain/CPF.java
/*
 * Copyright (c) 2019.
 * Fábrica de Software - Instituto de Informática (UFG)
 * Creative Commons Attribution 4.0 International License.
 */

package com.github.alicefng.cs.aula1.domain;

/**
 * Implementação de algoritomos para validar CPF
 */
public class CPF {

    /**
     * Avalia se uma String contém apenas dígitos
     *
     * @param cpf A string contendo o CPF a ser avaliado
     * @return verdadeiro, se a string contém apenos dígitos
     * falso, se a string contém outros caracteres
     */
    private static boolean sequenciaContemApenasDigitos(final String cpf) {
        return cpf.chars().allMatch(c -> Character.isDigit(c));
    }

    /**
     * Transforma caracteres numéricos em inteiros
     *
     * @param cpf String que contém CPF
     * @return Vetor contendo em cada posição um dígito do CPF como inteiro
     */
    private static int[] converteCaracteresEmInteiros(String cpf) {
        return cpf.chars().map(c -> Character.getNumericValue(c)).toArray();
    }

    /**
     * Avalia se CPF fornecido é válido dado seus dígitos verificados (dois últimos dígitos) - algoritmo 1
     *
     * @param cpf String que contém CPF
     * @return verdadeiro, se dígitos verificadores do CPF estiverem corretos, falso, se não
     * @throws IllegalArgumentException Se argumento é null
     * @throws IllegalArgumentException Se argumento não contém 11 dígitos
     * @throws IllegalArgumentException Se argumento contém algo diferente de dígitos (0 a 9)
     */
    public static boolean validarDigitosCPF(String cpf) {
        if (cpf == null) {
            throw new IllegalArgumentException("argumento é null");
        }
        if (cpf.length() != 11) {
            throw new IllegalArgumentException("CPF deve ter 11 dígitos: " + cpf);
        }
        if (!sequenciaContemApenasDigitos(cpf)) {
            throw new IllegalArgumentException("CPF deve conter somente dígitos (0 a 9): " + cpf);
        }

        int[] digitosCpf = converteCaracteresEmInteiros(cpf);
        int calculoParcialDigito10 = digitosCpf[0];
        int calculoParcialDigito11 = digitosCpf[1];

        for (int i = 1; i < 9; i++) {
            calculoParcialDigito10 += digitosCpf[i] * (i + 1);
        }
        for (int i = 2; i < 10; i++) {
            calculoParcialDigito11 += digitosCpf[i] * i;
        }
        int calculoFinalDigito10 = (calculoParcialDigito10 % 11) % 10;
        int calculoFinalDigito11 = (calculoParcialDigito11 % 11) % 10;

        return calculoFinalDigito10 == digitosCpf[9] & calculoFinalDigito11 == digitosCpf[10];
    }

    /**
     * Avalia se CPF fornecido é válido dado seus dígitos verificados (dois últimos dígitos) - algoritmo 2
     *
     * @param cpf String que contém CPF
     * @return verdadeiro, se dígitos verificadores do CPF estiverem corretos,
     * falso, se não
     * @throws IllegalArgumentException Se argumento é null
     * @throws IllegalArgumentException Se argumento não contém 11 dígitos
     * @throws IllegalArgumentException Se argumento contém algo diferente de dígitos (0 a 9)
     */
    public static boolean validarDigitosCPF2(String cpf) {
        if (cpf == null) {
            throw new IllegalArgumentException("argumento é null");
        }
        if (cpf.length() != 11) {
            throw new IllegalArgumentException("CPF deve ter 11 dígitos: " + cpf);
        }
        if (!sequenciaContemApenasDigitos(cpf)) {
            throw new IllegalArgumentException("CPF deve conter somente dígitos (0 a 9): " + cpf);
        }

        int[] digitosCpf = converteCaracteresEmInteiros(cpf);
        int calculoParcialDigito11 = digitosCpf[8];
        int calculoParcialDigito10 = calculoParcialDigito11;

        for (int i = 7; i >= 0; i--) {
            calculoParcialDigito11 += digitosCpf[i];
            calculoParcialDigito10 += calculoParcialDigito11;
        }
        int calculoFinalDigito10 = (calculoParcialDigito10 % 11) % 10;
        int calculoFinalDigito11 = ((calculoParcialDigito10 - calculoParcialDigito11 + 9 * digitosCpf[9]) % 11) % 10;

        return calculoFinalDigito10 == digitosCpf[9] & calculoFinalDigito11 == digitosCpf[10];
    }

}

File: aula-01/js/funcoes/CPF.js
/**
 * Funções para verificar validade de CPFs
 */

/**
 * Transforma caracteres numéricos em inteiros
 *
 * @param {String} cpf String que contém CPF
 *
 * @returns {Array} Vetor contendo em cada posição um dígito do CPF como inteiro
 */
function converteCaracteresEmInteiros(cpf) {
    const digitosCpf = [];
    for (let i = 0; i < cpf.length; i++) {
        digitosCpf[i] = parseInt(cpf.charAt(i));
    }

    return digitosCpf;
}

/**
 * Avalia se CPF fornecido é válido dado seus dígitos verificados
 * (dois últimos dígitos)
 *
 * @param {String} cpf String que contém CPF
 *
 * @returns {boolean} Verdadeiro, se dígitos verificadores do CPF estiverem
 * corretos, Falso, se não
 *
 * @throws {RangeError} Se cpf tem mais que 11 caracteres
 * @throws {TypeError} Se cpf é null ou undefined
 */
function validarDigitosCPF(cpf) {
    if (cpf === null || cpf === undefined) {
        throw new TypeError("Argumento null ou undefined");
    }
    if (cpf.length !== 11) {
        throw new RangeError("CPF deve ter 11 dígitos: " + cpf);
    }

    const digitosCpf = converteCaracteresEmInteiros(cpf);
    let calculoParcialDigito10 = digitosCpf[0];
    let calculoParcialDigito11 = digitosCpf[1];

    for (let i = 1; i < 9; i++) {
        calculoParcialDigito10 = calculoParcialDigito10 + digitosCpf[i] * (i + 1);
    }

    for (let i = 2; i < 10; i++) {
        calculoParcialDigito11 = calculoParcialDigito11 + digitosCpf[i] * i;
    }

    const calculoFinalDigito10 = (calculoParcialDigito10 % 11) % 10;
    const calculoFinalDigito11 = (calculoParcialDigito11 % 11) % 10;

    return calculoFinalDigito10 === digitosCpf[9]
        && calculoFinalDigito11 === digitosCpf[10];
}

/**
 * Avalia se CPF fornecido é válido dado seus dígitos verificados
 * (dois últimos dígitos)
 *
 * @param {String} cpf String que contém CPF
 *
 * @returns {boolean} Verdadeiro, se dígitos verificadores do CPF estiverem
 * corretos, Falso, se não
 *
 * @throws {RangeError} Se cpf tem mais que 11 dígitos
 * @throws {TypeError} Se cpf é null ou undefined
 */
function validarDigitosCPF2(cpf) {
    if (cpf === null || cpf === undefined) {
        throw new TypeError("Argumento null ou undefined");
    }
    if (cpf.length !== 11) {
        throw new RangeError("CPF deve ter 11 dígitos: " + cpf);
    }

    const digitosCpf = converteCaracteresEmInteiros(cpf);
    let calculoParcialDigito11 = digitosCpf[8];
    let calculoParcialDigito10 = calculoParcialDigito11;
    for (let c = 7; c >= 0; c--) {
        calculoParcialDigito11 = calculoParcialDigito11 + digitosCpf[c];
        calculoParcialDigito10 = calculoParcialDigito10 + calculoParcialDigito11;
    }
    const calculoFinalDigito10 = (calculoParcialDigito10 % 11) % 10;
    const calculoFinalDigito11 = ((calculoParcialDigito10 - calculoParcialDigito11
        + 9 * digitosCpf[9]) % 11) % 10;

    return calculoFinalDigito10 === digitosCpf[9]
        && calculoFinalDigito11 === digitosCpf[10];
}

module.exports = {
    converteCaracteresEmInteiros,
    validarDigitosCPF,
    validarDigitosCPF2
};
